### PR TITLE
Perf improvement on change/lazy items rendering

### PR DESCRIPTION
**Pull request by lbebber**
Improves the performance of removing items and changing their quantities
Removes items groups and adds the lazy loading functionality on a per-item basis. Also makes sure that the key for each item persists, in order to avoid unwanted re-rendering.

Before:

https://user-images.githubusercontent.com/5691711/103376495-896e9e80-4abb-11eb-8a9e-97ee8ddcc24a.mp4

After:

https://user-images.githubusercontent.com/5691711/103376506-91c6d980-4abb-11eb-9aa4-e463619d1c72.mp4

Linked over on https://lbebber--carrefourbrfood.myvtex.com/cart

Add a billion items on the cart using the following link:

https://lbebber--carrefourbrfood.myvtex.com/checkout/cart/add/?sku=3970&qty=8&seller=1&sc=2&sku=5278&qty=4&seller=1&sc=2&sku=4345&qty=4&seller=1&sc=2&sku=2406&qty=2&seller=1&sc=2&sku=2967&qty=3&seller=1&sc=2&sku=5144&qty=4&seller=1&sc=2&sku=4533&qty=2&seller=1&sc=2&sku=4533&qty=1&seller=1&sc=2&sku=8509&qty=2&seller=1&sc=2&sku=9276&qty=3&seller=1&sc=2&sku=11060&qty=1&seller=1&sc=2&sku=5019&qty=3&seller=1&sc=2&sku=16095&qty=4&seller=1&sc=2&sku=5651&qty=2&seller=1&sc=2&sku=141&qty=3&seller=1&sc=2&sku=4718&qty=1&seller=1&sc=2&sku=2396&qty=2&seller=1&sc=2&sku=2396&qty=2&seller=1&sc=2&sku=8418&qty=6&seller=1&sc=2&sku=2102&qty=1&seller=1&sc=2&sku=2102&qty=1&seller=1&sc=2&sku=19660&qty=2&seller=1&sc=2&sku=19662&qty=1&seller=1&sc=2&sku=3464&qty=1&seller=1&sc=2&sku=3381&qty=4&seller=1&sc=2&sku=6143&qty=2&seller=1&sc=2&sku=6146&qty=2&seller=1&sc=2&sku=12132&qty=2&seller=1&sc=2&sku=6100&qty=2&seller=1&sc=2&sku=11714&qty=1&seller=1&sc=2&sku=6623&qty=1&seller=1&sc=2&sku=11689&qty=1&seller=1&sc=2&sku=16455&qty=2&seller=1&sc=2&sku=14520&qty=1&seller=1&sc=2&sku=12014&qty=1&seller=1&sc=2&sku=11628&qty=1&seller=1&sc=2&sku=11993&qty=1&seller=1&sc=2&sku=13256&qty=1&seller=1&sc=2&sku=11752&qty=1&seller=1&sc=2&sku=6535&qty=1&seller=1&sc=2&sku=11220&qty=1&seller=1&sc=2&sku=4510&qty=1&seller=1&sc=2&sku=8978&qty=1&seller=1&sc=2&sku=12573&qty=1&seller=1&sc=2&sku=18260&qty=3&seller=1&sc=2&sku=12552&qty=1&seller=1&sc=2&sku=12477&qty=1&seller=1&sc=2&sku=2510&qty=3&seller=1&sc=2&sku=15987&qty=2&seller=1&sc=2&sku=18255&qty=1&seller=1&sc=2&sku=18594&qty=1&seller=1&sc=2&sku=14925&qty=1&seller=1&sc=2&sku=18285&qty=1&seller=1&sc=2&sku=17106&qty=1&seller=1&sc=2&sku=15955&qty=1&seller=1&sc=2&sku=16531&qty=1&seller=1&sc=2&sku=7269&qty=1&seller=1&sc=2&sku=1248&qty=1&seller=1&sc=2&sku=8987&qty=1&seller=1&sc=2&sku=15011&qty=1&seller=1&sc=2&sku=4415&qty=2&seller=1&sc=2&sku=4411&qty=1&seller=1&sc=2&sku=4401&qty=1&seller=1&sc=2&sku=125&qty=1&seller=1&sc=2&sku=4870&qty=1&seller=1&sc=2&sku=15643&qty=1&seller=1&sc=2&sku=12921&qty=2&seller=1&sc=2&sku=4399&qty=1&seller=1&sc=2&sku=15856&qty=1&seller=1&sc=2&sku=164&qty=1&seller=1&sc=2&sku=6991&qty=1&seller=1&sc=2&sku=2519&qty=1&seller=1&sc=2&sku=15858&qty=1&seller=1&sc=2&sku=4450&qty=1&seller=1&sc=2&sku=14778&qty=1&seller=1&sc=2&sku=4446&qty=1&seller=1&sc=2&sku=2594&qty=1&seller=1&sc=2&sku=14964&qty=1&seller=1&sc=2&sku=15231&qty=2&seller=1&sc=2&sku=2026&qty=1&seller=1&sc=2&sku=6804&qty=1&seller=1&sc=2&sku=2175&qty=3&seller=1&sc=2&sku=6385&qty=1&seller=1&sc=2&sku=4698&qty=1&seller=1&sc=2&sku=3053&qty=2&seller=1&sc=2&sku=14073&qty=1&seller=1&sc=2&sku=12076&qty=2&seller=1&sc=2&sku=17690&qty=4&seller=1&sc=2&sku=14371&qty=1&seller=1&sc=2&sku=2842&qty=1&seller=1&sc=2&sku=5777&qty=1&seller=1&sc=2&sku=1035&qty=1&seller=1&sc=2&sku=9350&qty=3&seller=1&sc=2&sku=3100&qty=6&seller=1&sc=2&sku=3100&qty=5&seller=1&sc=2

### CODE REVIEW

File: react/ProductList.tsx [hunk 1]
```diff
@@ -185,13 +154,22 @@ const ProductList: React.FC<Props> = (props) => {
...
-          items={group}
-        />
-      ))}
+      {unavailableItems.map((item) => { 
+        const { userType, allowManualPrice, children } = props
+        const shouldAllowManualPrice = allowManualPrice && userType === CALL_CENTER_OPERATOR
+
+        return (
+          <ItemContextWrapper
+            key={`${item.uniqueId}-${item.sellingPrice}`}
```
Comment (lbebber) [inline react/ProductList.tsx:174]: this is important

Comment (lbebber) [reply react/ProductList.tsx:174]: (removing the `index` from the key, if compared to how the component was being inserted on the `ProductGroup` component)

File: react/ProductList.tsx [hunk 2]
```diff
@@ -204,13 +182,21 @@ const ProductList: React.FC<Props> = (props) => {
...
-          {...props}
-          items={group}
-        />
-      ))}
+      {availableItems.map(item => {
+        const { userType, allowManualPrice, children } = props
+        const shouldAllowManualPrice = allowManualPrice && userType === CALL_CENTER_OPERATOR
+        return (
+          <ItemContextWrapper
+            key={`${item.uniqueId}-${item.sellingPrice}`}
```
Comment (lbebber) [inline react/ProductList.tsx:205]: idem

File: react/LazyRender.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import React, { FunctionComponent, useRef, useState } from 'react'
```
Comment (lbebber) [inline react/LazyRender.tsx]: This just wraps the lazy rendering hook into a component for easier handling.

copied from https://github.com/vtex-apps/render-runtime/blob/master/react/components/LazyRender.tsx

**Review by alvim — APPROVED** [review]
Nice!

There isn't much left to do. I reviewed the use of uniqueId in the key prop, because I was suspicious about some cases where the items could have the same uniqueId and prices, like items with different attachment values. But that will never happen, because items with different attachment values also have different uniqueId's actually.

File: react/LazyRender.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import { useOnView } from './hooks/useOnView'
```
Comment (klzns) [inline react/LazyRender.tsx]: Maybe we can add this component to the app
https://github.com/vtex-apps/on-view

Comment (alvim) [reply react/LazyRender.tsx]: Nice. I'll do it.

File: react/__tests__/ProductList.test.tsx [hunk 5]
```diff
@@ -1,3 +1,5 @@
+jest.mock("../LazyRender", () => ({ children } = { children: undefined }) => children);
```
Comment (lucasecdb) [inline react/__tests__/ProductList.test.tsx]: this needs to be below the imports

File: react/ProductList.tsx [hunk 6]
```diff
@@ -85,39 +83,12 @@ const ItemContextWrapper = memo<ItemWrapperProps>(function ItemContextWrapper({
...
-          item={item}
-          itemIndex={item.index}
-          shouldAllowManualPrice={shouldAllowManualPrice}
-          {...props}
-        >
-          {children}
-        </ItemContextWrapper>
-      ))}
-    </>
+    <LazyRender height={100} offset={300}>
```
Comment (lucasecdb) [inline react/ProductList.tsx]: any chance we can make this into a prop so agencies can customize the product height?

**Review by lucasecdb — APPROVED** [review]
left a comment, but LGTM

File: react/ProductList.tsx (hunk 6 above)
Comment (alvim) [reply react/ProductList.tsx]: I thinks it's a good suggestion. I'll review it.

**Review by lucasecdb — APPROVED** [review]

**Pull request merged**